scripts/src/java/org/oppia/android/scripts/coverage/CoverageReporter.kt

```diff
@@ -0,0 +1,198 @@
+package org.oppia.android.scripts.coverage
+
+import org.oppia.android.scripts.proto.Coverage
+import org.oppia.android.scripts.proto.CoverageReport
+import java.io.File
+
+/**
+ * Class responsible for generating rich text coverage report.
+ *
+ * @param repoRoot the root directory of the repository
+ * @param coverageReportList the list of coverage data proto
+ * @param reportFormat the format in which the report will be generated
+ */
+class CoverageReporter(
+  private val repoRoot: String,
+  private val coverageReportList: List<CoverageReport>,
+  private val reportFormat: ReportFormat,
+) {
+  private val computedCoverageRatio = computeCoverageRatio()
+  private val formattedCoveragePercentage = "%.2f".format(computedCoverageRatio * 100)
+
+  private val filePath = coverageReportList.firstOrNull()?.filePath ?: "Unknown"
+
+  private val totalLinesFound = coverageReportList.getOrNull(0)?.linesFound ?: 0
+  private val totalLinesHit = coverageReportList.getOrNull(0)?.linesHit ?: 0
+
+  /**
+   * Generates a rich text report for the analysed coverage data based on the specified format.
+   * It supports Markdown and HTML formats.
+   *
+   * @return a pair where the first value is the computed coverage ratio represented in [0, 1]
+   *     and the second value is the generated report text
+   */
+  fun generateRichTextReport(): Pair<Float, String> {
+    println("report format: $reportFormat")
+    return when (reportFormat) {
+      ReportFormat.MARKDOWN -> generateMarkdownReport()
+      ReportFormat.HTML -> generateHtmlReport()
+    }
+  }
+
+  private fun generateMarkdownReport(): Pair<Float, String> {
```

Comment:
My concern with the MD report is that unlike the html report, it does not show any information about which lines were not covered, so that it does not tell the developer what to do to next. Is that by design?

Reply:
Yeah @adhiamboperes, it makes sense that it looks like lacking details, but the plan (as far as I understood) was MD reports are majorly for ci runs to provide a short summary of the coverage that will be uploaded as GitHub comments (in ci which would be multiple files) and HTML reports for detailed coverage.

And during the meetings Ben suggested the templates for how these should look like:

> Any template for report generation?
> HTML: directory structure that matches the files under test, and then copies code over with red/green/yellow highlighting based on coverage type (probably can’t do, so just red/green for miss/hit).
> MD: should focus on a summarization of the test results, probably something like a cascading bullet list representing the covered tests and their percentages and covered/total lines of code, so something like:
> 
> Total coverage:
> Files covered: (# changed / # run with coverage)
> Coverage percentage: ##% covered / ##% expected
> LOC: # covered / # instrumented
> 
> (indent left) Specific coverage:
> app/src/.../app
> home
> HomeActivity.kt - 87% (110/115)
> ...
> ...
> Note: the path abbreviation is a good nice-to-have, but not required. Want to reduce cascading. Nice-to-have also would be adding a Markdown link to the file, e.g. for HomeActivity.kt above.

_(also the provided template it for M2 (when we have multiple files to check coverage for))_

and this was followed along, but if that's absolutely needed and could be useful maybe we can add it later as per feedback?

Reply:
Ok, it makes sense that we plan on expanding the details when we have more files/data